src/classes/RD2_OpportunityEvaluationService.cls

```diff
@@ -100,21 +118,41 @@ public class RD2_OpportunityEvaluationService {
 
         Map<npe03__Recurring_Donation__c, List<Date>> rdsWithCloseDates = new Map<npe03__Recurring_Donation__c, List<Date>>();
 
+        List<npe03__Recurring_Donation__c> updatedRDs = new List<npe03__Recurring_Donation__c>();
+
         for (npe03__Recurring_Donation__c rd : rds) {
-            if (!isCandidate(rd)) {
+
+            // Ignore Closed RD's with no active schedule
+            if (!isEligibleForEval(rd)) {
                 continue;
             }
 
+            // TODO Recalculate NextDonationDate before doing anything
+
+            // Collect the list of Opps that require a new Installment Opp
             if (isNewRecurringDonation(rd)) {
                 rdsWithCloseDates.put(rd, getInstallmentOppCloseDates(rd) );
                 continue;
             }
+
+            if (rd.Status__c == RD2_Constants.STATUS_CLOSED && hasOpenInstallments(rd)) {
+                // TODO Handle Open Installments on Closed Opp
+            }
+
+            // TODO Schedule Visualizer
             // TODO Update the Future Installment Opportunity if needed
+            // TODO Handle Past Due Installment
+
         }
 
-        // TODO Recalculate NextDonationDate
-        // Date nextDonationDate = scheduleService.calculateNextDonationDate(rd);
+        if (disableRollupsBeforeDml) {
+            TDTM_ProcessControl.toggleTriggerState('Opportunity', 'CRLP_Rollup_TDTM', false);
+            TDTM_ProcessControl.toggleTriggerState('npe01__OppPayment__c', 'CRLP_Rollup_TDTM', false);
+            TDTM_ProcessControl.toggleTriggerState('Allocation__c', 'CRLP_Rollup_TDTM', false);
+        }
 
+        // Create the next installment opportunity where needed
+        // TODO Refactor this to return a set of Opps to be inserted as opposed to having it do the DML in the service
```

Comment:
Do we have a WI that will handle this TODO?

Reply:
It's forthcoming.

Reply:
Just to add to this, the current OpportunityService commits the new Opps in the service method instead of returning instances of the new Opps. Though it does work in its current state, I believe that if we want to be able to better manage DML (performance) in the batch (actually the whole EvaluateOpportunities logic), the service should act more like the schedule or NDD services that can operate without actually committing the records.  What do you think?